### PR TITLE
All missing model functions added

### PR DESCRIPTION
**Pull request by Kajol-Kumari**
#34 [Feat.] add getConference() query to Conference model in Notifier-Service
#45 [Feat.] add getConferenceFromCategory() query to Conference model in Notifier-Service 
#46 [Feat.] add getCategories() query to Conference model in Notifier-Service
Functions added:
1. getCategories()
2. getConferences()
3. getConferencesFromCategory()

Test result of getConferences:
Screenshot from 2020-03-22 01-47-09

Test result of getConferencesFromCategory:
Screenshot from 2020-03-22 01-45-56

Test result of getCategories:
Screenshot from 2020-03-22 13-30-49

File changed:
src/models/conference.ts

### CODE REVIEW

File: Notifier-service/src/models/conference.ts [hunk 1]
```diff
@@ -1,65 +1,65 @@
...
-    private connection:Promise<Connection>;
+    private model: Promise<mongoose.Model<ConferenceDocument, {}>>;
+    private connection: Promise<Connection>;
     private logger = new Logger(this.constructor.name).getLogger();
-    constructor(database:Database){
+    constructor(database: Database) {
         super(database)
-        
+
         // this.connection = connection
```
Comment (rajatkb) [inline Notifier-service/src/models/conference.ts]: * clean up the comments , 😅 I might have left the when adding the makequery.

File: Notifier-service/src/models/conference.ts [hunk 2]
```diff
@@ -71,33 +71,55 @@ export class ConferenceModelI extends ConferenceModel{
...
-    async getConferences(offset:number , range:number):Promise<ConferenceDocument[] | null> {
-        let result = this.makeQuery((model) => {
-            return new Promise<ConferenceDocument[] | null>( (resolve , reject) => {
-                model.find({}).limit(range).exec((err , res) => {
-                    if(!err){
+    async getConferences(offset: number, range: number): Promise<ConferenceDocument[] | null> {
+
+        let result = this.model.then(model => {
+            return new Promise<ConferenceDocument[] | null>((resolve, reject) => {
+                model.find({ "deadline": { $gte: new Date() } }, (err, res) => {
```
Comment (rajatkb) [inline Notifier-service/src/models/conference.ts]: * use exec , find().sort().skip().limit().exec( callback )  , read about exec.

File: Notifier-service/src/models/conference.ts [hunk 3]
```diff
@@ -71,33 +71,55 @@ export class ConferenceModelI extends ConferenceModel{
...
         TO-DO
     */
-    async getConferences(offset:number , range:number):Promise<ConferenceDocument[] | null> {
-        let result = this.makeQuery((model) => {
-            return new Promise<ConferenceDocument[] | null>( (resolve , reject) => {
-                model.find({}).limit(range).exec((err , res) => {
-                    if(!err){
+    async getConferences(offset: number, range: number): Promise<ConferenceDocument[] | null> {
+
+        let result = this.model.then(model => {
```
Comment (rajatkb) [inline Notifier-service/src/models/conference.ts]: * Use makeQuery function ... the bove function has the logger activity and error handling already written in

File: Notifier-service/src/models/conference.ts [hunk 4]
```diff
@@ -71,33 +71,55 @@ export class ConferenceModelI extends ConferenceModel{
...
-        return result;
+        return result
     }
 
-    /*
-        TO-DO
-    */
-    async getConferencesFromCategory(category:string , offset:number , range:number):Promise<ConferenceDocument[] | null> {
-        return Promise.resolve(null);
+    async getConferencesFromCategory(category: string, offset: number, range: number): Promise<ConferenceDocument[] | null> {
```
Comment (rajatkb) [inline Notifier-service/src/models/conference.ts]: * Make similar changes as suggested above

File: Notifier-service/src/models/conference.ts [hunk 5]
```diff
@@ -71,33 +71,55 @@ export class ConferenceModelI extends ConferenceModel{
...
+                    }).skip(offset).limit(range);
+                })
+            })
+        return result
     }
 
-
-    async getCategories():Promise<Array<string> | null> {
-        return Promise.resolve(null);
+    async getCategories(): Promise<any> {
```
Comment (rajatkb) [inline Notifier-service/src/models/conference.ts]: * similar changes as above

File: Notifier-service/src/models/conference.ts [hunk 6]
```diff
@@ -68,50 +64,56 @@ export class ConferenceModelI extends ConferenceModel{
...
+        let result = this.makeQuery((model) => {
+                return new Promise<ConferenceDocument[] | null>((resolve, reject) => {
+                    model.find({ 'categories': { $in: [category] } }, (err, res) => {
+                        if (!err) {
+                            resolve(res);
+                        }
+                        else {
+                            reject(err);
+                        }
+                    }).skip(offset).limit(range);
```
Comment (rajatkb) [inline Notifier-service/src/models/conference.ts]: * No changes here ? Suggested to change it as above.

File: Notifier-service/src/models/conference.ts [hunk 7]
```diff
@@ -1,65 +1,61 @@
...
         return this.model
-        .then(callback)
-        .catch(error => {
-            this.logger.debug("Failed at getOne: model not initialised error:"+error);
-            this.logger.error("Failed at getOne : model must have failed to initialize :"+error);
-            return Promise.reject(new Error("model failed to be initialised"));
-        });                
+            .then(callback)
+            .catch(error => {
+                this.logger.debug("Failed at getOne: model not initialised error:" + error);
```
Comment (rajatkb) [inline Notifier-service/src/models/conference.ts]: * Change the error logger to reflect which function has failed. 
* "Failed at getOne"  should change to "Failed at --callbackname--"

File: Notifier-service/src/models/conference.ts (hunk 6 above)
Comment (Kajol-Kumari) [reply Notifier-service/src/models/conference.ts]: should i replace this one also with exec()?

File: Notifier-service/src/models/conference.ts [hunk 8]
```diff
@@ -68,50 +64,56 @@ export class ConferenceModelI extends ConferenceModel{
...
 
-    /*
-        TO-DO
-    */
-    async getConferencesFromCategory(category:string , offset:number , range:number):Promise<ConferenceDocument[] | null> {
-        return Promise.resolve(null);
+    async getConferencesFromCategory(category: string, offset: number, range: number): Promise<ConferenceDocument[] | null> {
+        let result = this.makeQuery((model) => {
+                return new Promise<ConferenceDocument[] | null>((resolve, reject) => {
+                    model.find({ 'categories': { $in: [category] } }).skip(offset).limit(range).exec((err, res) => {
```
Comment (rajatkb) [inline Notifier-service/src/models/conference.ts]: * sort missing like above.

**Pull request merged**